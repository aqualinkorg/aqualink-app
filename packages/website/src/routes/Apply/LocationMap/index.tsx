import React from "react";
import { Map, TileLayer, Marker } from "react-leaflet";
import L from "leaflet";
import { withStyles, WithStyles, createStyles, Theme } from "@material-ui/core";

import marker from "../../../assets/marker.png";

const pinIcon = L.icon({
  iconUrl: marker,
  iconSize: [20, 30],
  iconAnchor: [10, 30],
  popupAnchor: [0, -41],
});

const LocationMap = ({
  markerPositionLat,
  markerPositionLng,
  updateMarkerPosition,
  classes,
}: LocationMapProps) => {
  function updateLatLng(event: L.LeafletMouseEvent) {
    const { lat, lng } = event.latlng;
    updateMarkerPosition([lat, lng]);
  }

  function parseCoordinates(coord: string, defaultValue: number) {
    const parsed = parseFloat(coord);
    return Number.isNaN(parsed) ? defaultValue : parsed;
  }

  const markerPosition: L.LatLngTuple = [
    parseCoordinates(markerPositionLat, 37.773972),
    parseCoordinates(markerPositionLng, -122.431297),
  ];

  return (
    <Map
      center={markerPosition}
      zoom={5}
      className={classes.map}
      onclick={updateLatLng}
    >
      <TileLayer url="https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}" />
      {markerPosition && <Marker icon={pinIcon} position={markerPosition} />}
    </Map>
  );
};

const styles = (theme: Theme) => {
  return createStyles({
    map: {
      height: "100%",
      width: "100%",
      borderRadius: 4,
      cursor: "pointer",

      [theme.breakpoints.down("md")]: {
        height: 400,
      },
    },
  });
};

export interface LocationMapProps extends WithStyles<typeof styles> {
  markerPositionLat: string;
  markerPositionLng: string;
  updateMarkerPosition(tuple: L.LatLngTuple): void;
}

export default withStyles(styles)(LocationMap);
