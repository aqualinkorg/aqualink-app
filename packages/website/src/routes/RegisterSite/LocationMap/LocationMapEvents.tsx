import { LeafletEventHandlerFn } from 'leaflet';
import { useMapEvents } from 'react-leaflet';

type LocationMapEventsProps = {
  updateMarkerPosition: (position: [number, number]) => void;
  onZoomEnd: LeafletEventHandlerFn;
};

export const LocationMapEvents = ({
  updateMarkerPosition,
  onZoomEnd,
}: LocationMapEventsProps) => {
  useMapEvents({
    click(event) {
      const { lat, lng } = event.latlng.wrap();
      updateMarkerPosition([lat, lng]);
    },
    zoomend: onZoomEnd,
  });
  return null;
};
