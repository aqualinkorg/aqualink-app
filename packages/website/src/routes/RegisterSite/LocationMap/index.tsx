import React, { useCallback, useState } from 'react';
import { MapContainer, TileLayer, Marker, useMapEvents } from 'react-leaflet';
import L, { LeafletEvent } from 'leaflet';
import { Theme } from '@mui/material';

import { WithStyles } from '@mui/styles';
import withStyles from '@mui/styles/withStyles';
import createStyles from '@mui/styles/createStyles';

import { mapConstants } from 'constants/maps';
import marker from '../../../assets/marker.png';

const pinIcon = L.icon({
  iconUrl: marker,
  iconSize: [20, 30],
  iconAnchor: [10, 30],
  popupAnchor: [0, -41],
});

const INITIAL_ZOOM = 5;

const MapEventsHandler = ({
  updateLatLng,
  setZoom,
}: {
  updateLatLng: (event: L.LeafletMouseEvent) => void;
  setZoom: (event: LeafletEvent) => void;
}) => {
  useMapEvents({
    click: updateLatLng,
    zoomend: setZoom,
  });
  return null;
};

const LocationMap = ({
  markerPositionLat,
  markerPositionLng,
  updateMarkerPosition,
  classes,
}: LocationMapProps) => {
  const [zoom, setZoom] = useState<number>(INITIAL_ZOOM);

  const onZoomEnd = useCallback(
    // eslint-disable-next-line no-underscore-dangle
    (event: LeafletEvent) => setZoom(event.target._zoom as number),
    [],
  );

  function updateLatLng(event: L.LeafletMouseEvent) {
    const { lat, lng } = event.latlng.wrap();
    updateMarkerPosition([lat, lng]);
  }

  function parseCoordinates(coord: string, defaultValue: number) {
    const parsed = parseFloat(coord);
    return Number.isNaN(parsed) ? defaultValue : parsed;
  }

  const markerPosition: L.LatLngTuple = [
    parseCoordinates(markerPositionLat, 37.773972),
    parseCoordinates(markerPositionLng, -122.431297),
  ];

  return (
    <MapContainer
      center={markerPosition}
      zoom={zoom}
      className={classes.map}
      maxBounds={mapConstants.MAX_BOUNDS}
      maxBoundsViscosity={1.0}
      minZoom={1}
    >
      <TileLayer url="https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}" />
      {markerPosition && <Marker icon={pinIcon} position={markerPosition} />}
      <MapEventsHandler
        updateLatLng={(e) => updateLatLng(e)}
        setZoom={(e) => onZoomEnd(e)}
      />
    </MapContainer>
  );
};

const styles = (theme: Theme) => {
  return createStyles({
    map: {
      height: '100%',
      width: '100%',
      borderRadius: 4,
      cursor: 'pointer',

      [theme.breakpoints.down('lg')]: {
        height: 400,
      },
    },
  });
};

export interface LocationMapProps extends WithStyles<typeof styles> {
  markerPositionLat: string;
  markerPositionLng: string;
  updateMarkerPosition(tuple: L.LatLngTuple): void;
}

export default withStyles(styles)(LocationMap);
