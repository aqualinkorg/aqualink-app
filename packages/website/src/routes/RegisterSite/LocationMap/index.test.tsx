import React from "react";
import { render } from "@testing-library/react";
import LocationMap from ".";

test("renders as expected", () => {
  const { container } = render(
    <LocationMap
      markerPositionLat="1"
      markerPositionLng="1"
      updateMarkerPosition={() => {}}
    />
  );
  expect(container).toMatchSnapshot();
});
