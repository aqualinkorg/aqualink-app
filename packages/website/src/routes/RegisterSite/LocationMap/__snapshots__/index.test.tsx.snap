// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    class="LocationMap-map-1 leaflet-container leaflet-grab leaflet-touch-drag"
    style="position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      >
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 18; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/5/15/16"
              style="width: 256px; height: 256px; left: -23px; top: -233px;"
            />
          </div>
        </div>
      </div>
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      >
        <img
          alt=""
          class="leaflet-marker-icon leaflet-zoom-hide leaflet-interactive"
          src="marker.png"
          style="margin-left: -10px; margin-top: -30px; width: 20px; height: 30px; left: 0px; top: 0px; z-index: 0;"
          tabindex="0"
        />
      </div>
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            +
          </a>
          <a
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            âˆ’
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JS library for interactive maps"
          >
            Leaflet
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
